lib/services/local/headlessauthn_watcher_test.go

```diff
@@ -26,171 +26,324 @@ import (
+		watcherClock  clockwork.FakeClock
+		watcherCancel context.CancelFunc
+		identity      *local.IdentityService
+		buf           *backend.CircularBuffer
+	}
 
-	watcherCtx, watcherCancel := context.WithCancel(ctx)
-	defer watcherCancel()
+	newSuite := func(t *testing.T) *testSuite {
+		identity := newIdentityService(t, clockwork.NewFakeClock())
 
-	watcherClock := clockwork.NewFakeClock()
-	w, err := local.NewHeadlessAuthenticationWatcher(watcherCtx, local.HeadlessAuthenticationWatcherConfig{
-		Clock:   watcherClock,
-		Backend: identity.Backend,
-	})
-	require.NoError(t, err)
+		// use a standalone buffer as a watcher service.
+		buf := backend.NewCircularBuffer()
+		buf.SetInit()
 
-	pubUUID := services.NewHeadlessAuthenticationID([]byte(sshPubKey))
+		watcherCtx, watcherCancel := context.WithCancel(ctx)
+		t.Cleanup(func() {
+			watcherCancel()
+		})
+
+		watcherClock := clockwork.NewFakeClock()
+		w, err := local.NewHeadlessAuthenticationWatcher(watcherCtx, local.HeadlessAuthenticationWatcherConfig{
+			Clock:          watcherClock,
+			WatcherService: buf,
+			Backend:        identity.Backend,
+		})
+		require.NoError(t, err)
 
-	waitInGoroutine := func(ctx context.Context, t *testing.T, name string, cond func(*types.HeadlessAuthentication) (bool, error)) (chan *types.HeadlessAuthentication, chan error) {
+		return &testSuite{
+			watcher:       w,
+			watcherClock:  watcherClock,
+			watcherCancel: watcherCancel,
+			identity:      identity,
+			buf:           buf,
+		}
+	}
+
+	waitInGoroutine := func(ctx context.Context, t *testing.T, watcher *local.HeadlessAuthenticationWatcher, name string, cond func(*types.HeadlessAuthentication) (bool, error)) (chan *types.HeadlessAuthentication, chan error) {
 		headlessAuthnCh := make(chan *types.HeadlessAuthentication, 1)
 		errC := make(chan error, 1)
 		go func() {
-			headlessAuthn, err := w.Wait(ctx, name, cond)
+			headlessAuthn, err := watcher.Wait(ctx, name, cond)
 			errC <- err
 			headlessAuthnCh <- headlessAuthn
 		}()
-		require.Eventually(t, func() bool { return w.CheckWaiter(name) }, time.Millisecond*100, time.Millisecond*10)
-
 		return headlessAuthnCh, errC
 	}
 
-	t.Run("WaitTimeout", func(t *testing.T) {
-		waitCtx, waitCancel := context.WithTimeout(ctx, time.Millisecond*10)
+	t.Run("WaitEventWithConditionMet", func(t *testing.T) {
+		t.Parallel()
+		s := newSuite(t)
+
+		waitCtx, waitCancel := context.WithTimeout(ctx, time.Second*5)
 		defer waitCancel()
 
-		_, err = w.Wait(waitCtx, pubUUID, func(ha *types.HeadlessAuthentication) (bool, error) { return true, nil })
-		require.Error(t, err)
-		require.Equal(t, waitCtx.Err(), err)
+		firstEmitReceived := make(chan struct{})
+		headlessAuthnCh, errC := waitInGoroutine(waitCtx, t, s.watcher, pubUUID, func(ha *types.HeadlessAuthentication) (bool, error) {
+			select {
+			case <-firstEmitReceived:
+			default:
+				close(firstEmitReceived)
+			}
+
+			return ha.User != "", nil
+		})
```

Comment:
Could we refactor this block and the require.Eventually below into something we can reuse? They appear together a few times.

Eg:

```suggestion
		eventuallyFirstEvent := waitForFirstEmit(waitCtx, t, s.watcher)
		
		// ...
		
		// then, instead of the require.Eventually below:
		eventuallyFirstEvent()
````